tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+    auto decryptorBuilder = AESCipherBuilder{operationMode, _secretKey}.setIV(iv);
+    std::shared_ptr<SymmetricCipher> decryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        decryptor = encryptoBuilder.buildAuthenticatedDecryptor();
+    }
+    else {
+        decryptor = encryptoBuilder.buildDecryptor();
+    }
+
+    std::vector<uint8_t> decryptedText;
+
+    // Decrypt packets and read decrypted text in chunks. Size of these chunks is different.
+    static int const DecryptionChunkSize = 1024 * 4;
+    static int const ReaderBufferSize = 1024;
+    auto ciphertextRunner = ciphertext.begin();
+    while (std::distance(ciphertextRunner, std::end(ciphertext)) > DecryptionChunkSize) {
+        decryptor->update({ciphertextRunner, ciphertextRunner += DecryptionChunkSize});
+        auto decryptedChunk = decryptor->read(ReaderBufferSize);
+        decryptedText.insert(std::end(decryptedText), std::begin(decryptedChunk), std::end(decryptedChunk));
+    }
+    // Decrypt the rest which didn't fit into decryption chunk
+    decryptor->update({ciphertextRunner, std::end(ciphertext)});
+
+    auto authenticatedDecryptor = dynamic_cast<AuthenticatedEncryptionI*>(decryptor.get());
+    if (authenticatedDecryptor) {
+        authenticatedDecryptor->setAuthTag(tag);
+    }
+    auto decryptedChunkFinal = decryptor->finish();
+    decryptedText.insert(std::end(decryptedText), std::begin(decryptedChunkFinal), std::end(decryptedChunkFinal));
+
+    ASSERT_THAT(decryptedText, ::testing::ElementsAreArray(_plaintext));
+}
+
+INSTANTIATE_TEST_SUITE_P(Chunks,
+                         SymmetricCipherAdvancedTest,
+                         testing::ValuesIn(AllSupportedCIpherModesToTest));
+
+
+
+
+class SymmetricCipherMemoryStrategy : public  ::testing::Test {
+protected:
+    void SetUp() override {
+        for (int i = 0; i < kNumberOfTestBlocks; i++) {
+            std::vector<uint8_t> block(kTestBlockLength, i);
+            _expectedData.insert(std::end(_expectedData), std::begin(block), std::end(block));
+        }
+    }
+
+    QueueOfVectorsMemoryStrategy sut;
+
+    const int kNumberOfTestBlocks = 4;
+    const int kTestBlockLength = 16;
+    std::vector<uint8_t > _expectedData;
+};
+
+TEST_F(SymmetricCipherMemoryStrategy, SeveralWritesAndReadAll) {
+    sut.write({std::begin(_expectedData), std::begin(_expectedData) + 3});
+    sut.write({std::begin(_expectedData) + 3, std::begin(_expectedData) + 7});
+    sut.write({std::begin(_expectedData) + 7, std::end(_expectedData)});
+
+    auto readChunk = sut.readAll();
+    ASSERT_THAT(readChunk, ::testing::ElementsAreArray(_expectedData));
+}
+
+TEST_F(SymmetricCipherMemoryStrategy, EmptyReadAll) {
+    auto readBlock = sut.read(kTestBlockLength);
+    ASSERT_TRUE(readBlock.empty());
+}
+
+TEST_F(SymmetricCipherMemoryStrategy, ReadInWrittenBlockSizes) {
+    auto dataPacketizer = std::begin(_expectedData);
+
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
```

Comment:
a bit less dense code would be more readable here imho

Reply:
I have refactored the assertion out. It should be a bit cleaner.